Add verbosity option to `install tool` command.

peterhuene opened the pull request:
This PR adds the `--verbosity` option to the `install tool` command.

MSBuild/NuGet output is now controllable by the user and defaults to being "quiet".

This enables users to see warnings from NuGet that otherwise would be swallowed
unless NuGet returned a non-zero exit code. As a byproduct of this change, the
exception handling and error messages related to obtaining tool packages was
retooled. We no longer display `install tool` command line help for installation
failures, as it should only be displayed for command line syntax errors.

Fixes #8465.

wli3 commented on src/dotnet/ToolPackage/IProjectRestorer.cs:
```diff
@@ -11,6 +12,7 @@ internal interface IProjectRestorer
             FilePath? nugetconfig,
-            string source);
+            string source,
+            IEnumerable<string> restoreArguments);
```
Either merge all the param to restoreArguments including source, nuget config. Or explicit about what restore arguments allow to pass.

peterhuene replied:
Yeah, I was looking for a better way to do this.  The problem is that I didn't want to have to define a `verbosity` option into this command because it is currently implemented as a "forwarded argument".  Thus I chose to simply pass through the forwarded arguments this way rather than taking a single option for "verbosity".  I'd be happy to do it the other way and just define the option manually rather than using `CommonOptions.VerbosityOption` if you think that's best.

wli3 replied:
I think "CommonOptions.VerbosityOption" is better. restoreArguments feel like the caller can passing a lot of opinions. But a string array cannot compose

wli3 commented on src/dotnet/commands/dotnet-install/dotnet-install-tool/InstallToolCommand.cs:
```diff
@@ -67,69 +70,68 @@ internal class InstallToolCommand : CommandBase
-                    string.Format(LocalizableStrings.InstallFailedNuget,
-                        ex.Message),
-                    innerException: ex);
+                errorReporter.WriteLine(ex.Message);
```
Now, this class is directly handing low level output and format of the error message?

peterhuene replied:
Yes, it should.  Throwing `GracefulException` should only be reserved for command line syntax errors as it purposely shows the help from the tool; that indicates to users they just need to "type something different to solve the problem.  For failures installing the tool or in the tool itself (such as configuration), the error should be logged by the command.

I should make this `.Red()` though like the base command uses for errors.

wli3 replied:
".Red()" -- I haven't dug about the color issue too. I remember there is a problem if the user set the command line profile differently (like change back ground to white). I don't know if we just pic a red and it will work for all profile.

I don't know if we have a abstraction already. What we really want is a presentation to call attention.

wli3 commented on test/dotnet-install-tool.Tests/GivenDotnetInstallTool.cs:
```diff
@@ -0,0 +1,46 @@
+        [Fact]
+        public void ItErrorsWithANonExistentSource()
+        {
+            var result = new InstallCommand()
```
This will trigger adding PATH to CI machine.

And to use a very high level test(slow and flaky) to test error message does not worth it.

It is a indication of no good separation of error handing ("format what color what goes to print log vs error" with rest of the code like download, parsing, path handling)

wli3 replied:
We could have one or two such high level test to **double check** basic scenarios. But high level test should be reduced much as possible.

wli3 replied:
We actually have a good separation in production code.
You can just pass in a action => () = throw TheExceptionYouWant to installcommand, and assert on IReport

peterhuene replied:
Can you elaborate on what `trigger adding PATH to CI machine` means?

This particular test is intended to make sure the `source` option is being passed down through the various layers and not the error message itself; just the fact it errors means the source option was passed through correctly.  I think this test is poorly named and should be renamed at the very least.

This test, however, would have caught the missing line in the command passing in the source option with my original PR.

Additionally, because we're spawning another process in these tests, we can easily capture the output and examine it; this is desired for testing an option like `verbose` which changes MSBuild output.

wli3 replied:
> Can you elaborate on what trigger adding PATH to CI machine means?

1. During first run experience, CLI will try to add .dotnet/tools to PATH. But it will not when certain environment variables are set.

2. During dotnet install, it will check and print messages.

And dependends on how to assert, due to .dotnet/tools is not present on PATH, test may behavir differently.

> This particular test is intended to make sure the source option is being passed down through the various layers and not the error message itself; just the fact it errors means the source option was passed through correctly. I think this test is poorly named and should be renamed at the very least.

> This test, however, would have caught the missing line in the command passing in the source option with my original PR.

This test gaps is fixed by https://github.com/dotnet/cli/pull/8402 . It is unfortunate it takes so long. But this should be in the same PR as InstallCommand.cs

peterhuene commented:
I'll address these concerns (and fix the test failure).

wli3 commented on test/dotnet-install-tool.Tests/GivenDotnetInstallTool.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        [Fact]
+        public void ItRunsWithTheSpecifiedVerbosity()
```
This ok, verbosity depends on nuget and msbuild behavior, to get enough confident, we need to boot up nuget anyway. The assertion is too fragile, what if nuget change behavior. it is a contract nuget verbose:n will print out the package id? What if nuget will echo package id all the time?

`* put on perfectionist hat`
There is test needed:
1. verbosity test in adapter layer. Pass down to nuget as /v:n, or nuget will behave differently.
2. mock test, to cover InstallToolCommand
3. convert --verb to correct method param and help text

This explicit three layer is awkward, maybe we should build something. When parser see `--verbose`, `--source`, it should pass `verbose`, "source feed" etc in "context" like async context, and it should be passed implicit everywhere we call msbuild or nuget. But it should also have a way to prevent other objects modify them along the way.

peterhuene commented:
@wli3 I've updated the PR.  I fixed passing a collection of "restore arguments" in favor of passing in a "verbosity" argument.  I've removed the questionable "source error" test since this was better tested already in the command tests.  There are now two high-level integration tests: one that tests that verbosity is quiet by default and one that tests verbosity can be overridden.  I kept them simple high-level tests.

peterhuene commented:
@KathleenDollard would you mind reviewing the .resx file changes for strings and let me know if there's any outstanding questions from the discussion on #8465.  Thanks!

KathleenDollard commented on src/dotnet/CommonLocalizableStrings.resx:
```diff
@@ -520,28 +520,26 @@
-    <value>The tool's settings file has more than one command defined.</value>
+  <data name="ToolSettingsMissingEntryPoint" xml:space="preserve">
+    <value>Tool entry point is missing.</value>
   </data>
```
I would love to find better language than "entry point" but I can't come up with any.

tannergooding replied:
The C# compiler uses the following if you attempt to compile an exe without a Main method:
`Program does not contain a static 'Main' method suitable for an entry point`

tannergooding replied:
The runtime gives the following when attempting to execute a PE file without an entry point:
`Entry point not found in assembly.`

peterhuene replied:
In this context, it is that the tool defines a command with a missing `EntryPoint` (i.e. the assembly to execute) setting.  This would only happen if users published a tool with a malformed settings XML file.

This should really be `Command 'bar' is missing an entry point setting.` since it is for a particular command (currently only one is supporter per tool).

The resulting error would currently be:

```
The settings file in the tool's NuGet package is invalid: Tool entry point is missing.
Tool 'foo' failed to install. Please contact the tool author for assistance.
```

peterhuene replied:
I've changed the message to my recent suggestion.

KathleenDollard commented:
@peterhuene 

I am slightly confused by the description. 

In the default mode, we are using verbosity of quiet?
This results in output if and only if the return code from NuGet is nonzero?
If the user sets the verbosity to "normal" it outputs warning from NuGet?

If you can confirm those things, I'm happy to approve. If not, we should perhaps chat.

peterhuene commented:
Before this change:

`dotnet install tool` would run the restore operation with the default verbosity (I believe this is actually `minimal`) and buffer the redirected output.  The output would only be displayed to the user when the restore operation failed.  This meant that any warnings coming from a successful restore operation were being suppressed entirely.

With this change:

`dotnet install tool` runs the restore operation with a user-specified verbosity (defaulting to `quiet`) and lets the child process inherit the stdout/stderr handles, thus warnings and errors are directly displayed to the user regardless of the restore operation failing or succeeding.

With respect to the restore operation, the difference between `minimal` and `quiet` verbosity is that the `minimal` level prints out messages like:

```
Restoring packages for foo.csproj
...
Restore completed in 32.7 ms for foo.csproj
```

Effectively, using `quiet` hides the fact a restore operation is taking place unless warnings/errors are produced.
